### PR TITLE
[WIP] Introducing metadata collections in dav

### PR DESCRIPTION
**Pull request by sharidas**
This change is to have metadata collection
in dav app. This will help users to:
1) Create fileid with a UUID
2) Create subfolder "p" inside "meta" folder
3) Create subfolder "v" inside "meta" folder
The "p" folder stands for preview and "v"
folder stands for version.

Signed-off-by: Sujith H <sharidasan@owncloud.com>

## Description
Lets focus on versions for now

## Todo
- [ ] write behat tests to test dav api @Peter-Prochaska 

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: apps/dav/lib/MetaData/MetaDataHome.php [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+	private $principalInfo;
+
+	private $metaData;
+
+	/**
+	 * MetaDataHome constructor.
+	 *
+	 * @param array $principalInfo
+	 */
+	public function __construct($principalInfo, IMetaData $metaData) {
```
Comment (DeepDiver1975) [inline apps/dav/lib/MetaData/MetaDataHome.php]: there should be no principalInfo here - this collection is not user based

File: apps/dav/lib/MetaData/RootCollection.php [hunk 2]
```diff
@@ -0,0 +1,29 @@
+<?php
+
+namespace OCA\DAV\MetaData;
+
+use Sabre\DAVACL\AbstractPrincipalCollection;
+use Sabre\DAVACL\IPrincipal;
+
+class RootCollection extends AbstractPrincipalCollection {
```
Comment (DeepDiver1975) [inline apps/dav/lib/MetaData/RootCollection.php]: no principal collection - there is no user involved.

we want 'meta/$fileid/v' not 'meta/$user/$fileId/v

File: apps/dav/lib/MetaData/RootCollection.php [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+	 *
+	 * The passed array contains principal information, and is guaranteed to
+	 * at least contain a uri item. Other properties may or may not be
+	 * supplied by the authentication backend.
+	 *
+	 * @param array $principalInfo
+	 * @return MetaDataHome
+	 */
+	function getChildForPrincipal(array $principalInfo) {
+		$metaData = \OC::$server->getMetaData();
```
Comment (DeepDiver1975) [inline apps/dav/lib/MetaData/RootCollection.php]: Please have a look at #29207 

```php
$metaNodeOfFile = \OC::$server->getRootFolder()->get("meta/{$info->getId()}/v");

**Review by DeepDiver1975 — CHANGES_REQUESTED** [review]

File: lib/public/IMetaData.php [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+ *
+ */
+
+namespace OCP;
+use OCP\Files\File;
+use OCP\Files\NotFoundException;
+
+/**
+ * This class provides avatar functionality
+ * @since 6.0.0
```
Comment (PVince81) [inline lib/public/IMetaData.php]: 10.0.4 ?

File: lib/public/IMetaData.php [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>
+ *
+ */
+
+namespace OCP;
+use OCP\Files\File;
+use OCP\Files\NotFoundException;
+
+/**
+ * This class provides avatar functionality
```
Comment (PVince81) [inline lib/public/IMetaData.php]: not avatar

File: lib/public/IMetaData.php (hunk 4 above)
Comment (DeepDiver1975) [reply lib/public/IMetaData.php]: this class will go away ... dont worry

File: apps/dav/lib/SystemTag/SystemTagsByIdCollection.php [hunk 6]
```diff
@@ -149,7 +149,7 @@ function delete() {
 	}
 
 	function getName() {
-		return 'systemtags';
+		return 'systemtags poda thendi';
```
Comment (DeepDiver1975) [inline apps/dav/lib/SystemTag/SystemTagsByIdCollection.php]: ???

Comment (sharidas) [reply apps/dav/lib/SystemTag/SystemTagsByIdCollection.php]: Removed. Got added to PR by mistake.

File: lib/private/Files/Storage/Local.php [hunk 7]
```diff
@@ -441,4 +441,25 @@ public function moveFromStorage(\OCP\Files\Storage $sourceStorage, $sourceIntern
 			return parent::moveFromStorage($sourceStorage, $sourceInternalPath, $targetInternalPath);
 		}
 	}
+
```
Comment (DeepDiver1975) [inline lib/private/Files/Storage/Local.php]: TODO:
- [ ] try to move this to class Common

File: apps/dav/lib/MetaData/MetaDataHome.php [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+use Sabre\DAV\Exception\MethodNotAllowed;
+use Sabre\DAV\Exception\NotFound;
+use Sabre\DAV\ICollection;
+use Sabre\HTTP\URLUtil;
+
+class MetaDataHome implements ICollection {
+
+	/** @var array */
+
+	private $metaData;
```
Comment (DeepDiver1975) [inline apps/dav/lib/MetaData/MetaDataHome.php]: unused

File: apps/dav/lib/MetaData/MetaDataHome.php [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+
+	function createFile($name, $data = null) {
+		throw new Forbidden('Permission denied to create a file');
+	}
+
+	function createDirectory($name) {
+		throw new Forbidden('Permission denied to create a folder');
+	}
+
+	function getChild($name) {
```
Comment (DeepDiver1975) [inline apps/dav/lib/MetaData/MetaDataHome.php]: name ignored?

File: apps/dav/lib/MetaData/MetaDataNode.php [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+	 */
+	public function __construct(MetaVersionCollection $metaVersionCollection) {
+		$this->metaVersionCollection = $metaVersionCollection;
+		$this->nodeFolder = \OC::$server->getRootFolder()->get('');
+	}
+
+	/**
+	 * @return \OCP\Files\Node
+	 */
+	public function getMetaNodeFolder() {
```
Comment (DeepDiver1975) [inline apps/dav/lib/MetaData/MetaDataNode.php]: public?

File: apps/dav/lib/MetaData/MetaDataNode.php [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+	 */
+	public function getMetaPreviewFolder() {
+		return \OC::$server->getRootFolder()->get('meta/'. $this->metaVersionCollection->getId() . '/p');
+	}
+
+	/**
+	 *  Create folder meta
+	 */
+	public function createMetaNode() {
+		$this->nodeFolder->getStorage()->mkdir('meta');
```
Comment (DeepDiver1975) [inline apps/dav/lib/MetaData/MetaDataNode.php]: as discussed - meta sub tree is read only

**Review by DeepDiver1975 — CHANGES_REQUESTED** [review]

**Comment by DeepDiver1975** [comment]
closing this in favor of #29281

**Pull request closed without merging**